Mathlib/LinearAlgebra/IsTensorProduct.lean

```diff
@@ -0,0 +1,362 @@
+variable {N : Type uN} [AddCommMonoid N] [Module R N]
+variable {X : Type uX} [AddCommMonoid X]
+
+variable (M N X) in
+structure BalancedBiAddMonoidHom extends (M →+ N →+ X) where
+(balanced : ∀ (r : R) (m : M) (n : N), toAddMonoidHom (op r • m) n = toAddMonoidHom m (r • n))
+
+instance BalancedBiAddMonoidHom.funLike :
+    FunLike (BalancedBiAddMonoidHom R M N X) M (fun _ => N → X) where
+  coe f := fun m => fun n => f.toAddMonoidHom m n
+  coe_injective' f g h := by
+    cases f
+    cases g
+    congr
+    exact AddMonoidHom.ext fun m => AddMonoidHom.ext fun n => congr_fun (congr_fun h m) n
+
+instance BalancedBiAddMonoidHom.addMonoidHomClass :
+    AddMonoidHomClass (BalancedBiAddMonoidHom R M N X) M (N →+ X) where
+  coe f m := f.toAddMonoidHom m
+  coe_injective' f g h := FunLike.ext _ _ fun m => funext fun n =>
+    FunLike.congr_fun (congr_fun h m) n
+  map_add f _ _ := FunLike.ext _ _ fun _ => by
+    exact FunLike.congr_fun (f.toAddMonoidHom.map_add _ _) _
+  map_zero f := FunLike.ext _ _ fun _ => by
+    exact FunLike.congr_fun (f.toAddMonoidHom.map_zero) _
+
+@[simp] lemma BalancedBiAddMonoidHom.apply_apply_zero (f : BalancedBiAddMonoidHom R M N X) (m : M) :
+    f m 0 = 0 :=
+  (f.toAddMonoidHom m).map_zero
+
+lemma BalancedBiAddMonoidHom.apply_apply_add (f : BalancedBiAddMonoidHom R M N X)
+    (m : M) (n n' : N):
+    f m (n + n') = f m n + f m n' :=
+  (f.toAddMonoidHom m).map_add n n'
+
+@[simp] lemma BalancedBiAddMonoidHom.apply_zero_apply (f : BalancedBiAddMonoidHom R M N X) (n : N) :
+    f 0 n = 0 :=
+  FunLike.congr_fun f.toAddMonoidHom.map_zero _
+
+lemma BalancedBiAddMonoidHom.apply_add_apply (f : BalancedBiAddMonoidHom R M N X)
+    (m m' : M) (n : N):
+    f (m + m') n = f m n + f m' n :=
+  FunLike.congr_fun (f.toAddMonoidHom.map_add m m') n
+
+lemma BalancedBiAddMonoidHom.apply_smul_apply (f : BalancedBiAddMonoidHom R M N X)
+    (r : R) (m : M) (n : N) :
+    f (op r • m) n = f m (r • n) :=
+  f.balanced r m n
+
+class IsTensorProduct (tmul : BalancedBiAddMonoidHom R M N X) :=
+(lift ⦃P : Type uX⦄ [AddCommMonoid P] (_ : BalancedBiAddMonoidHom R M N P) : X →+ P)
+(lift_comp ⦃P : Type uX⦄ [AddCommMonoid P] (f : BalancedBiAddMonoidHom R M N P) (m : M) (n : N) :
+  lift f (tmul m n) = f m n)
+(lift_uniq ⦃P : Type uX⦄ [AddCommMonoid P] (f : BalancedBiAddMonoidHom R M N P) (fHat : X →+ P) :
+  (∀ m n, fHat (tmul m n) = f m n) → lift f = fHat)
+
+end defs
+
+section addGroups
+-- I feel like these should work for "semimodules" as well, but I don't know how to prove it
+open IsTensorProduct
+
+universe uR uM uN uX
+
+variable (R : Type uR) [Semiring R]
+variable {M : Type uM} [AddCommGroup M] [Module Rᵐᵒᵖ M]
+variable {N : Type uN} [AddCommGroup N] [Module R N]
+variable {X : Type uX} [AddCommGroup X]
+
+variable (tmul : BalancedBiAddMonoidHom R M N X) [IsTensorProduct R tmul]
+
+lemma IsTensorProduct.mem_closure (x : X) :
+    x ∈ AddSubgroup.closure { y | ∃ (m : M) (n : N), y = tmul m n } := by
```

Comment:
You may be able to prove the [AddSubmonoid.closure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/Submonoid/Basic.html#AddSubmonoid.closure) version by first showing that for any two AddCommMonoid satisfying the IsTensorProduct predicate is isomorphic (in a way compatible with the tmul), and carry out the explicit [TensorProduct](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct.html#TensorProduct)[TensorProduct](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct.html#TensorProduct) construction, which clearly is generated by the `tmul m n`. The only thing you need to change is [TensorProduct.Eqv.of_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct.html#TensorProduct.Eqv), I believe.

Reply:
Kevin reminded me that one can use the characteristic predicate to prove `X -> span` and `span -> X` compose to identity on `span` and `X`, so that this must be true. Thanks for your suggestion